### PR TITLE
Add retry_if_exception_message

### PR DESCRIPTION
**Pull request by Brian-Williams**
From issue #125.

Add retry classes for exception messages.

I'm very open to name suggestions. Specifically wondering if `retry_if_not_exception_message` or `retry_until_exception_message` is best.

`retry_if_exception_message` and it's complement act different out of necessity. The normal class can end if there is no exception whereas it's inverse needs to retry if there is no exception message.

It's worth noting in my branch unit tests are intermittently failing on [wait class backwards incompatibility](https://travis-ci.org/Brian-Williams/tenacity/jobs/405236070#L470). I don't believe this is caused by any of my changes, but it's worth noting.

### CODE REVIEW

File: tenacity/retry.py [hunk 1]
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
...
+            prog = re.compile(match)
+
+            def match_fnc(exception_message):
+                return prog.match(str(exception_message))
+            return match_fnc
+        raise TypeError(
+            "{}() missing 1 required argument 'message' or 'match'".format(
+                self.__class__.__name__))
+
+    def _no_failure(self):
```
Comment (jd) [inline tenacity/retry.py]: `@staticmethod`

Not sure it's worth having a method, just an attribute would be enough.

File: tenacity/retry.py [hunk 2]
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
...
+        predicate = super(
+            retry_if_not_exception_message, self
+        )._predicate_factory(*args, **kwargs)
+
+        def inverted_predicate(*args_, **kwargs_):
+            return not predicate(*args_, **kwargs_)
+
+        return inverted_predicate
+
+    def _no_failure(self):
```
Comment (jd) [inline tenacity/retry.py]: @staticmethod

Not sure it's worth having a method, just an attribute would be enough.

File: tenacity/retry.py [hunk 3]
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
...
+class retry_if_exception_message(retry_base):
+    """Retries if an exception message equals or matches."""
+
+    def __init__(self, message=None, match=None, require_exception=False):
+        if message and match:
+            raise TypeError(
+                "{}() takes either 'message' or 'match', not both".format(
+                    self.__class__.__name__))
+        self.predicate = self._predicate_factory(message, match)
+        self.require_exception = require_exception
```
Comment (jd) [inline tenacity/retry.py]: That does not seem to be related to the heart of the function. I'd remove that, and put that in a different class (and PR) if that's needed.

File: tenacity/retry.py [hunk 4]
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
             return not self.predicate(attempt.result())
 
 
+class retry_if_exception_message(retry_base):
```
Comment (jd) [inline tenacity/retry.py]: you should inherit from `retry_if_exception`

File: tenacity/retry.py [hunk 5]
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
...
+            return match_fnc
+        raise TypeError(
+            "{}() missing 1 required argument 'message' or 'match'".format(
+                self.__class__.__name__))
+
+    def _no_failure(self):
+        """Retry logic for no failure."""
+        return False
+
+    def __call__(self, attempt):
```
Comment (jd) [inline tenacity/retry.py]: This is not needed if you inherit from `retry_if_exception`

File: tenacity/retry.py [hunk 6]
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
...
 
+class retry_if_exception_message(retry_base):
+    """Retries if an exception message equals or matches."""
+
+    def __init__(self, message=None, match=None, require_exception=False):
+        if message and match:
+            raise TypeError(
+                "{}() takes either 'message' or 'match', not both".format(
+                    self.__class__.__name__))
+        self.predicate = self._predicate_factory(message, match)
```
Comment (jd) [inline tenacity/retry.py]: just pass the result to the `__init__` method of `retry_if_exception` here (since you will inherit from it). The factory method is not even really needed IMHO, you can put that code in `__init__`.

File: tenacity/retry.py [hunk 7]
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
...
+        if not attempt.failed:
+            if self.require_exception:
+                raise RuntimeError("No exception to get errormessage from")
+            return self._no_failure()
+        return self.predicate(attempt.exception().value)
+
+
+class retry_if_not_exception_message(retry_if_exception_message):
+    """Retries until an exception message equals or matches."""
+
```
Comment (jd) [inline tenacity/retry.py]: just set invert the predicate in `__init__` with a lambda I'd say.

**Comment by Brian-Williams** [comment]
Thanks for the review!

Removed `require_exception`, which allowed us to inherit from `retry_if_exception` as suggested.

Removed `_no_failure` fnc as it no longer has a use case if `retry_if_not_exception_message` is the only `__call__` in this PR; `retry_if_exception_message` can simply inherit that magic method.

Fixed a bug in the assumption of the errormessage retrieval and wrote an additional unit test to handle the positive case of `retry_if_not_exception_message`.

Rebased on jd/master to a single commit.
